# Facets of Mathematics: Week 6

## Aims

By the end of this notebook you should:

* Be familiar with lines and conics in the projective plane.
* Be able to test basic properties of conics in the projective plane, such as (ir)reducibility and smoothness.
* Be able to find a conic that contains $5$ general points in the projective plane.


## Topics

* The projective plane
* Lines in the projective plane
* Conics in projective plane
* Smooth and singular curves

Before attempting this notebook you should first have looked at the accompanying videos for week 6. 

This notebook combines explanations of the key week 6 concepts, with programming exercises which you should attempt by adding or editing code.

## References

**Algebriac geometry for sophomores:** I. Cheltsov, available from Learn
(Course Materials $\rightarrow$ Theme 2: Cubic Curves $\rightarrow$ Week 6 $\rightarrow$ Week 6 Tutorial)

* The projective plane: Section 2.a.
* Lines in the projective plane Section 2.b.
* Conics in the projective plane Sections 2.c., 2.e. and 2.f.
* Smooth and singular curves Section 2.h.


## 1. What is the projective plane?

To define the complex projective plane, we define an equivalence relation $\sim$ on the set $\mathbb{C}^{3}\setminus \{(0,0,0)\}$ as follows:

$$
\big(x,y,z\big)\sim \big(x^{\prime},y^{\prime},z^{\prime}\big)\iff \big(x,y,z\big)=\big(\lambda x^{\prime},\lambda y^{\prime},\lambda z^{\prime}\big)\ \text{for some complex number $\lambda\ne 0$}.
$$

For every $(x,y,z)\ne(0,0,0)$, we denote by $[x:y:z]$ the set consisting all points

$$
\big(\lambda x,\lambda y,\lambda z\big)
$$

where $\lambda$ runs through all non-zero complex numbers.
Then

$$
\big[x:y:z\big]=\big[x^{\prime}:y^{\prime}:z^{\prime}\big]\iff \big(x,y,z\big)\sim \big(x^{\prime},y^{\prime},z^{\prime}\big).
$$

For example, we have $[1:2:3]=[2:4:6]=[1+i:2+2i:3+3i]$.

### Definition
The complex projective plane $\mathbb{P}^{2}_{\mathbb{C}}$ is the set consisting of all *three-tuples*

$$
\big[x:y:z\big].
$$

Here $(x,y,z)$ is a point in $\mathbb{C}^3$ such that $(x,y,z)\ne (0,0,0)$.

For geometric reasons, we will refer to the elements of the set $\mathbb{P}^2_{\mathbb{C}}$ as points.
Let us repeat that  points in $\mathbb{P}^2_{\mathbb{C}}$ are *three-tuples* $[x:y:z]$ such that

$$
\big[x:y:z\big]=\big[x^{\prime}:y^{\prime}:z^{\prime}\big]\iff\ \big(x,y,z\big)=\big( \lambda x^{\prime}, \lambda y^{\prime},\lambda z^{\prime}\big)\ \text{for some non-zero}\ \lambda\in\mathbb{C}.
$$

Here we exclude the point $[0:0:0]$.

The plane $\mathbb{P}^2_{\mathbb{C}}$ consists of $\mathbb{C}^2$ and a *line at infinity*.
To be precise, let $U_z$ be the subset in $\mathbb{P}^2_{\mathbb{C}}$ consisting of points $[x:y:z]$ with $z\ne 0$.
For each point $[x:y:z]\in U_z$, we have

$$
\big[x:y:z\big]=\Big[\frac{x}{z}:\frac{y}{z}:1\Big],
$$

and the point $[x:y:z]$ is uniquely determined by the pair $(\frac{x}{z},\frac{y}{z})$.
Thus, we can identify $U_z$ and $\mathbb{C}^2$.
Let $\overline{x}=\frac{x}{z}$ and $\overline{y}=\frac{y}{z}$.
Then we can consider $\overline{x}$ and $\overline{y}$ as coordinates on $U_z$.

What is the complement in $\mathbb{P}^2_{\mathbb{C}}$ of the subset $U_z$?
The complement is the subset in $\mathbb{P}^2_{\mathbb{C}}$ consisting of all points $[x:y:0]$
where $x$ and $y$ are complex numbers such that $x\ne 0$ or $y\ne 0$ (or both).
Thus, we can identify the complement of $U_z$ in $\mathbb{P}^2_{\mathbb{C}}$ with the complex projective line $\mathbb{P}^1_{\mathbb{C}}$.
This is the *line at infinity*.

We defined $U_z$ as the subset in the plane $\mathbb{P}^2_{\mathbb{C}}$ that consists of all points $[x:y:z]$ with $z\ne 0$.
Then we identified $U_z$ and $\mathbb{C}^2$ with two coordinates $\overline{x}=\frac{x}{z}$ and $\overline{y}=\frac{y}{z}$.
Why did we choose $z$ here? There is no particular reason for this.
Similar to $U_z$, we can define $U_x$ and $U_y$.

## 2. Lines in the projective plane

To learn how to live in the complex projective plane,
we have to consider simple geometric objects in $\mathbb{P}^2_{\mathbb{C}}$ and play with them.
The simplest objects in $\mathbb{P}^2_{\mathbb{C}}$ are lines. How to define them?
We already know one line: the line given by $z=0$, which we called the *line at infinity*.
Motivated by this and our definition of a line in $\mathbb{C}^2$, define a line in $\mathbb{P}^2_{\mathbb{C}}$ as a subset given by

$$
Ax+By+Cz=0
$$

for some (fixed) complex numbers $A$, $B$ and $C$ such that $(A,B,C)\ne(0,0,0)$.
This subset does not change if we multiply the equation by a non-zero complex number $\lambda$,
since the equation

$$
\lambda Ax+\lambda By+\lambda Cz=0
$$

defines the same subset in $\mathbb{P}^2_{\mathbb{C}}$.
Thus, the line given by $Ax+By+Cz=0$ is determined by the point $[A:B:C]\in\mathbb{P}^2_{\mathbb{C}}$.

As in the complex plane $\mathbb{C}^2$, a line in $\mathbb{P}^2_{\mathbb{C}}$ is uniquely determined by any two distinct points contained within it.
To be precise, if $P$ and $Q$ are points in $\mathbb{P}^2_{\mathbb{C}}$ such that $P\ne Q$,
then there is a unique line in $\mathbb{P}^2_{\mathbb{C}}$ that contains both $P$ and $Q$.
This line is given by the following equation:

$$
 \mathrm{det}\left(
\begin{array}{ccc}
a & b & c \\
a^{\prime} & b^{\prime} & c^{\prime} \\
x & y & z \\
\end{array}\right)=0,
$$

where $P=[a:b:c]$ and $Q=[a^{\prime}:b^{\prime}:c^{\prime}]$.

For instance, to find the line in $\mathbb{P}^2_{\mathbb{C}}$ that contains points $[11:-7:1]$ and $[2:5:1]$,
we have to solve the system of linear equations

$$
\begin{aligned}
11A-7B+C &=0,\\
2A+5B+C &=0.
\end{aligned}
$$


The solutions of this system form a one-dimensional vector space by the rank-nullity theorem.
Using Python, we see that $(A,B,C)=(4,3,-23)$ is one solution, so that $(4,3,-23)$ is a basis of this vector space.
This means that the line $4x+3y-23z=0$ is the unique line in $\mathbb{P}^2_{\mathbb{C}}$ that contains the points $[11:-7:1]$ and $[2:5:1]$.

We can carry out these computations using Python:

In [1]:
from sympy import *

A, B, C, x , y, z = symbols('A, B, C, x, y, z')

L = A * x + B * y + C * z

P = (11, -7, 1)
Q = (2, 5, 1)

E1 = L.subs(((x, P[0]), (y, P[1]), (z, P[2])))
E2 = L.subs(((x, Q[0]), (y, Q[1]), (z, Q[2])))

solution = solve((E1, E2, C + 23), A, B, C)

coeff = (solution[A], solution[B], solution[C])

L = L.subs(((A, coeff[0]), (B, coeff[1]), (C, coeff[2])))

print(L)

4*x + 3*y - 23*z


Of course, we can find the same line using explicit determinant equation

$$
\mathrm{det}\left(
\begin{array}{ccc}
11 & -7 & 1 \\
2 & 5 & 1 \\
x & y & z \\
\end{array}\right)=0.
$$

Using Python, this can be done as follows:

In [2]:
from sympy import *

x, y, z = symbols('x, y, z')

P = (11, -7, 1)
Q = (2, 5, 1)

M = Matrix([P, Q, [x, y, z]])
L = M.det()

print(L)

-12*x - 9*y + 69*z


This gives us the line $-12x - 9y + 69z=0$, which is the same line as $4x+3y-23z=0$.

### Exercise 1

Find the line in $\mathbb{P}^2_{\mathbb{C}}$ containing $[1+3i:7-2i:1]$ and $[3:8-i:1]$.

In [21]:
from sympy import *

P = (1 + 3 * I, 7 - 2 * I, 1)
Q = (3, 8 - I, 1)

x, y, z = symbols('x, y, z')
M = Matrix([P, Q, [x, y, z]])

L = M.det()     # [x, y. z] is a set of lineal combinations of P and Q, hence determinatee of [x,y,z], P, Q is zero 

print(L, '= 0')

# or to sovle the linear equations 

a, b, c, x, y, z = symbols ('a, b, c, x, y, z')
L = a * x + b * y + c * z

E1 = L.subs(x, P[0]).subs(y, P[1]).subs(z, P[2])
E2 = L.subs(x, Q[0]).subs(y, Q[1]).subs(z, Q[2])

coef = solve((E1, E2, a + 1 + I  ), a, b, c)
print (coef)
L = L.subs(a,coef[a]).subs(b, coef[b]).subs(c, coef[c])
print(L, '= 0')
L = L.subs(a,expand_complex(coef[a])).subs(b, expand_complex(coef[b])).subs(c, expand_complex(coef[c]))
print(L, '= 0')

-x - I*x + 2*y - 3*I*y - 10*z + 29*I*z = 0
{a: -1 - I, b: 2 - 3*I, c: -10 + 29*I}
x*(-1 - I) + y*(2 - 3*I) + z*(-10 + 29*I) = 0
x*(-1 - I) + y*(2 - 3*I) + z*(-10 + 29*I) = 0


If three points $[a_1:b_1:c_1]$, $[a_2:b_2:c_2]$ and $[a_3:b_3:c_3]$ in $\mathbb{P}^2_{\mathbb{C}}$
are contained in a single line, then we say that they are *collinear*.
Using determinant equation of a line passing through two given points in  $\mathbb{P}^2_{\mathbb{C}}$,
we see that this happens if and only if

$$
\mathrm{det}\left(
\begin{array}{ccc}
a_1&b_1&c_1 \\
a_2&b_2&c_2 \\
a_3&b_3&c_3 \\
\end{array}\right)=0.
$$

### Exercise 2

For this exercise we consider the points

$$
P_0 := [0:0:1], \quad P_1 := [0:1:1], \quad P_2 := [1:1:-1], \quad P_3 := [1:3:1], \quad P_4 := [2:5:1],  \quad P_5 := [1:1:1], \quad P_6 := [1:4:2].
$$

<b>a)</b> Write a Python script to determine whether the points $P_0$, $P_1$, $P_2$ are colinear.

In [13]:
from sympy import *

P0 = (0, 0, 1)
P1 = (0, 1, 1)
P2 = (1, 1, -1)

M = Matrix([P0, P1, P2])
if M.det == 0:
    print ('They are colinear')
else:
    print('They are not colinear')
# add code here

They are not colinear


<b>b)</b> Consider the following Python script. Discuss with other members of your group what the code outputs and how each line works. Add comments to explain how the code works.

In [4]:
from sympy import *

# Define our points and collect them in a tuple

P0 = (0, 0, 1)
P1 = (0, 1, 1)
P2 = (1, 1, -1)
P3 = (1, 3, 1)
P4 = (2, 5, 1)
P5 = (1, 1, 1)
P6 = (1, 4, 2)

points = (P0, P1, P2, P3, P4, P5, P6)

# add comment here

count = 0

# add comment here

for l in range(len(points)):
    for m in range(l + 1, len(points)):
        for n in range(m + 1, len(points)):
            
            # add comment here
            
            M = Matrix([points[l], points[m], points[n]])
            if M.det() == 0:
                count = count + 1
                print(points[l], points[m], points[n], ' are colinear')

# add comment here
                
print('Number of colinear triples:', count)

(0, 0, 1) (1, 1, -1) (1, 1, 1)  are colinear
(0, 1, 1) (1, 1, -1) (1, 3, 1)  are colinear
(0, 1, 1) (1, 1, -1) (2, 5, 1)  are colinear
(0, 1, 1) (1, 1, -1) (1, 4, 2)  are colinear
(0, 1, 1) (1, 3, 1) (2, 5, 1)  are colinear
(0, 1, 1) (1, 3, 1) (1, 4, 2)  are colinear
(0, 1, 1) (2, 5, 1) (1, 4, 2)  are colinear
(1, 1, -1) (1, 3, 1) (2, 5, 1)  are colinear
(1, 1, -1) (1, 3, 1) (1, 4, 2)  are colinear
(1, 1, -1) (2, 5, 1) (1, 4, 2)  are colinear
(1, 3, 1) (2, 5, 1) (1, 4, 2)  are colinear
Number of colinear triples: 11


The above code is messy because it uses nested loops. We can clean this up using ```itertools.combinations``` from the ```itertools``` library.

<b>c)</b> Consider the following Python script, which carries out the same task as in part b). As before, discuss with other members of your group how each line works and add comments. 

In [3]:
from sympy import *
import itertools

# Define our points and collect them in a tuple

P0 = (0, 0, 1)
P1 = (0, 1, 1)
P2 = (1, 1, -1)
P3 = (1, 3, 1)
P4 = (2, 5, 1)
P5 = (1, 1, 1)
P6 = (1, 4, 2)

points = (P0, P1, P2, P3, P4, P5, P6)

# define a list to store colinear triples 
triples = []

# using cpmbination of itertools to check any three vectors if they are colinear

for point0, point1, point2 in itertools.combinations(points, 3):
    
    # put three vector in a matrix M
    
    M = Matrix([point0, point1, point2])
    
    # check if determinate of M is zero, if yes stor triple in the list triples 
    
    if M.det() == 0:
        
        print(point0, point1, point2, ' are collinear')
        
        triples.append((point0, point1, point2))

# print the length of list triples        
        
print('Number of colinear triples:', len(triples))

(0, 0, 1) (1, 1, -1) (1, 1, 1)  are collinear
(0, 1, 1) (1, 1, -1) (1, 3, 1)  are collinear
(0, 1, 1) (1, 1, -1) (2, 5, 1)  are collinear
(0, 1, 1) (1, 1, -1) (1, 4, 2)  are collinear
(0, 1, 1) (1, 3, 1) (2, 5, 1)  are collinear
(0, 1, 1) (1, 3, 1) (1, 4, 2)  are collinear
(0, 1, 1) (2, 5, 1) (1, 4, 2)  are collinear
(1, 1, -1) (1, 3, 1) (2, 5, 1)  are collinear
(1, 1, -1) (1, 3, 1) (1, 4, 2)  are collinear
(1, 1, -1) (2, 5, 1) (1, 4, 2)  are collinear
(1, 3, 1) (2, 5, 1) (1, 4, 2)  are collinear
Number of colinear triples: 11


Why do we need the projective plane?
Is $\mathbb{C}^2$ not good enough? In many instances, working over the projective plane $\mathbb{P}^2_{\mathbb{C}}$ is much more convenient that working over $\mathbb{C}^2$. The first evidence for this is given by:

### Proposition

Let $L$ and $L^{\prime}$ be two distinct lines in $\mathbb{P}^2_{\mathbb{C}}$.
Then $L\cap L^{\prime}$ consists of exactly one point in $\mathbb{P}^2_{\mathbb{C}}$.
In particular, the intersection $L\cap L^{\prime}$ is not empty.

Thus, there are no *parallel* lines in $\mathbb{P}^2_{\mathbb{C}}$:
every pair of lines in $\mathbb{P}^2_{\mathbb{C}}$ has a common point and this point is unique provided that the lines are distinct.
For example, to find the unique intersection point of the lines $4x-3y-17z=0$ and $7x+5y-11z=0$ in $\mathbb{P}^2_{\mathbb{C}}$,
we can use the determinant formula:

$$
\mathrm{det}\left(
\begin{array}{ccc}
4 & -3 & -17 \\
7 & 5 & -11 \\
\textbf{i} & \textbf{j} & \textbf{k} \\
\end{array}\right)=118\textbf{i}-75\textbf{j}+41\textbf{k},
$$

where $\textbf{i}=(1,0,0)$, $\textbf{j}=(0,1,0)$ and $\textbf{k}=(0,0,1)$.
It gives us the point $[118:-75:41]$. We can use Python to compute this determinant:

In [4]:
from sympy import *

i, j, k = symbols('i, j, k')

M = Matrix([[4, -3, -17], [7, 5, -11], [i, j, k]])

determinant = simplify(M.det())

print(determinant)

118*i - 75*j + 41*k


Alternatively, we can find the unique intersection point of the lines $4x-3y-17z=0$ and $7x+5y-11z=0$
by solving the following system of linear equations:

$$
\begin{aligned}
4x-3y-17z&=0,\\
7x+5y-11z&=0.
\end{aligned}
$$

This system has a unique solution $[x:y:z]\in\mathbb{P}^2_{\mathbb{C}}$, which can be found using Python again:

In [5]:
from sympy import *

x, y, z = symbols('x, y, z')

L1 = 4 * x - 3 * y - 17 * z
L2 = 7 * x + 5 * y - 11 * z

solution = solve((L1, L2, z - 1), x, y, z)
P = (solution[x], solution[y], solution[z])

print(P)

(118/41, -75/41, 1)


This gives us the same intersection point $[\frac{118}{41}:-\frac{75}{41}:1]=[118:-75:41]$.
Note that we used an extra condition $z=1$ in this Python code.
Why did we do this?
By definition, at least one number among $x$, $y$ and $z$ is not zero,
because there is no point $[0:0:0]$. Thus, if $z\ne 0$, then

$$
\big[x:y:z\big]=\Big[\frac{x}{z}:\frac{y}{z}:1\Big],
$$

so that we can assume that $z=1$.
If $z=0$, then either $x\ne 0$ or $y\ne 0$ (or both), and we can do a similar trick in these cases.
Here we guessed that $z\ne 0$, since $(x,y,z)=(0,0,0)$ is the only solution to the system of equations

$$
\begin{aligned}
4x-3y-17z &= 0,\\
7x+5y-11z &= 0,\\
z &= 0,
\end{aligned}
$$


and we want to avoid the $(0, 0, 0)$ solution when working over the projective plane.

### Exercise 3

Find the intersection point in $\mathbb{P}^2_{\mathbb{C}}$ of the line given by

$$
(-13+17i)x+(23+29i)y-(11-3i)z=0
$$

and the line given by $$(2-5i)x+(15+19i)y+(37+41i)z=0.$$

In [29]:
from sympy import *

i, j, k = symbols('i,j,k')
P=(-13+17*I, 23+29*I,-(11-3*I))
Q=((2-5*I),(15+19*I),(37+41*I))

M=Matrix([P,Q,[i,j,k]])
determinant = simplify(M.det())

print(determinant)

i*(-116 + 2180*I) + j*(1171 - 35*I) + k*(-709 + 65*I)


## 3. Conics in projective plane

The natural subsets to work with in the complex projective plane are given by *homogeneous* polynomial equations in $x$, $y$ and $z$. A polynomial is said to be *homogeneous* if all terms have the same degree. Thus, 

$$
f(x,y,z) = 3x^2y + 6xy^2 + 12z^3
$$

is homogeneous but 

$$
g(x,y,z) = 3x^2y + 6xy^2 + 12z^2
$$

is not homogeneous. 

The simplest case is when we have one equation:

$$
f_d(x,y,z)=0,
$$

where $f_d(x,y,z)$ is a homogeneous polynomial of degree $d\geqslant 1$. Such subsets are called *complex projective plane curves* of degree $d$ or simply *plane curves*.
Plane curves of degree $1$ are just lines in $\mathbb{P}^2_{\mathbb{C}}$.
Plane curves of degree $2$ are called *conics*, and plane projective curves of degree $3$ are called *cubic curves*.

If the polynomial $f_d(x,y,z)$ is *irreducible*, we say that the curve it defines is *irreducible*.
Otherwise, we say that the curve is *reducible*.
Lines are always irreducible. But conics can be reducible.
However, if a conic is reducible, then it is a union of two (not necessarily distinct) lines.

Let $\mathcal{C}$ be a conic in $\mathbb{P}^2_{\mathbb{C}}$. Then it is given by

$$
Ax^{2}+Bxy+Cy^{2}+Dxz+Eyz+Fz^2=0
$$

for some complex numbers $A$, $B$, $C$, $D$, $E$ and $F$ such that $(A,B,C,D,E,F)\ne(0,0,0,0,0,0)$.
We can rewrite this equation in the matrix form as:

$$
 \left(
\begin{array}{ccc}
x & y & z \\
\end{array}\right)\left( \begin{array}{ccc}
A & \frac{B}{2} & \frac{D}{2} \\
\frac{B}{2} & C & \frac{E}{2} \\
\frac{D}{2} & \frac{E}{2} & F \\
\end{array}\right)\left(\begin{array}{c}
x\\
y\\
z\\
\end{array}\right)=0.
$$

Denote the $3\times 3$ matrix in this equation by $M$.
Similar to the result in the complex plane, the conic $\mathcal{C}$ is irreducible if and only if $\mathrm{det}(M)\ne 0$.

Instead of the matrix $M$, it is often more handy to use the following matrix:

\begin{equation}
\label{equation:Hessian-conic}
\left(\begin{array}{ccc}
2A & B & D \\
B & 2C & E \\
D & E & 2F \\
\end{array}\right).
\end{equation}

Its determinant is $8\times\mathrm{det}(M)$, so we can use it to check the irreducibility of the conic $\mathcal{C}$.
This is slightly better than using matrix $M$, since we do not need to divide by $2$ anywhere.
This matrix can be found using the following Python function:

In [6]:
from sympy import *

x, y, z, A, B, C, D, E, F = symbols('x, y, z, A, B, C, D, E, F')

# Here we are going to define our own Python function (using def). 
# This function takes a polynomial f as an its input, and returns the associated matrix.

def conicMatrix(f):
    
# We extract the coefficients of the polynomial f using the coeff_monomial command

    a11 = 2 * f.coeff_monomial(x ** 2)
    a12 = f.coeff_monomial(x * y)
    a13 = f.coeff_monomial(x * z)
    a21 = a12
    a22 = 2 * f.coeff_monomial(y ** 2)
    a23 = f.coeff_monomial(y * z)
    a31 = a13
    a32 = a23
    a33 = 2 * f.coeff_monomial(z * z)
    
    M = Matrix([[a11, a12, a13], [a21, a22, a23], [a31, a32, a33]])
    
    return M

# We now test our Python function conicMatrix on a polynomial. 
# Note that we define f to be a polynomial in x, y, z using Poly. 


f = Poly(A * x ** 2 + B * x * y + C * y ** 2 + D * x * z + E * y * z + F * z ** 2, x, y, z)
M = conicMatrix(f)

print(simplify(M.det()))

8*A*C*F - 2*A*E**2 - 2*B**2*F + 2*B*D*E - 2*C*D**2


Here we need to define ```f``` using ```Poly```. If we try to run the code with just the expression

```
f = A * x ** 2 + B * x * y + C * y ** 2 + D * x * z + E * y * z + F * z ** 2
```

then it will return an error. This is because ```coeff_monomial``` must take a polynomial input, as defined using ```Poly```.




Let's consider an explicit example.
Let $\mathcal{C}$ be the conic in $\mathbb{P}^2_{\mathbb{C}}$ given by the equation

$$
(1+i)x^{2}-7xy+(3-2i)y^{2}+(2+5i)xz+(-1+i)yz+(-11+13i)z^2=0.
$$

Then $\mathcal{C}$ is irreducible conic, since

$$
\mathrm{det}\left(\begin{matrix}
2+2i&-7&2+5i\\
-7&6-4i&-1+i\\
2+5i&-1+i&-22+26i
\end{matrix}\right)=674-1000i\ne 0.
$$

This is checked by adding

```A = 1 + I
 B = -7
 C = 3 - 2 * I
 D = 2 + 5 * I
 E = -1 + I
 F = -11 + 13 * I```

to the previous Python script, in the line above the definition of ```f```. 

A slightly different approach is given noting the desired matrix corresponds to the *Hessian* matrix of second derivatives of $f$:

$$
\mathrm{Hess}(f)(x, y, z) := 
\left(
\begin{array}{ccc}
\frac{\partial^2 f}{\partial x^2}(x,y,z) & \frac{\partial^2 f}{\partial x \partial y}(x,y,z) & \frac{\partial^2 f}{\partial x \partial z}(x,y,z) \\
\frac{\partial^2 f}{\partial y \partial x}(x,y,z) & \frac{\partial^2 f}{\partial y^2}(x,y,z) & \frac{\partial^2 f}{\partial y \partial z}(x,y,z) \\
\frac{\partial^2 f}{\partial z \partial x}(x,y,z) & \frac{\partial^2 f}{\partial z \partial y}(x,y,z) & \frac{\partial^2 f}{\partial z^2}(x,y,z) 
\end{array}
\right).
$$

```SymPy``` has an inbuilt Hessian function, which allows us to compute the matrix in an efficient manner. 

In [7]:
from sympy import *

x, y, z = symbols('x, y, z')

# Define our conic

A = 1 + I
B = -7
C = 3 - 2 * I
D = 2 + 5 * I
E = -1 + I
F = -11 + 13 * I

f = A * x ** 2 + B * x * y + C * y ** 2 + D * x * z + E * y * z + F * z ** 2

# Compute the Hessian matrix

M = hessian(f, (x, y, z))

# Print the determinant of the Hessian

print(simplify(M.det()))

674 - 1000*I


Here the function ```hessian``` handles ```f``` an expression in the sense we are familiar with (whereas the ```conicMatrix``` function deals with polynomials as elements of the class ```Poly```). This is not a big difference, since we can always convert elements of the class ```Poly``` into expressions using Python function ```as_expr()```.


In [8]:
from sympy import *

x, y, z, A, B, C, D, E, F = symbols('x, y, z, A, B, C, D, E, F')

f = Poly(A * x ** 2 + B * x * y + C * y ** 2 + D * x * z + E * y * z + F * z ** 2, x, y, z)
g = f.as_expr()

print(g)

A*x**2 + B*x*y + C*y**2 + D*x*z + E*y*z + F*z**2


### Exercise 4

Let $\mathcal{C}$ be the conic in $\mathbb{P}^2_{\mathbb{C}}$ that is given by

$$
x^2+xy-2y^2+3xz+3yz+z^2=0.
$$

Show that the conic $\mathcal{C}$ is *irreducible*.
Verify that $\mathcal{C}$ contains the point $[-2:1:3]$.

In [38]:
from sympy import *

# Show that the conic  is irreducible. Verify that  contains the point [−2:1:3].

A = 1
B = 1
C = -2
D = 3
E = 3
F = 1

x, y, z = symbols('x, y, z')
f = A * x ** 2 + B * x * y + C * y ** 2 + D * x * z + E * y * z + F * z ** 2
# save curve as C5 for exercise 5
C5 = f
M = hessian(f, (x, y, z))
print(simplify(M.det()))

V = expand_complex(f.subs(x,-2).subs(y,1).subs(z,3))

print(V)

V = simplify(f.subs(x,-2).subs(y,1).subs(z,3))
print(V)

if V == 0:
    print('on the curve')

18
0
0
on the curve


A line and an irreducible conic in $\mathbb{P}^2_{\mathbb{C}}$ always meet.
Moreover, their intersection consists of at most $2$ points by [Cheltsov, Lemma 2.13].
In fact, it consists of **precisely** two points if we count them with appropriate multiplicities.
Furthermore, to find the intersection points of a line and a conic in $\mathbb{P}^2_{\mathbb{C}}$,
the only thing you need to know is how to solve a quadratic equation in one variable.


For example, let $L$ be the line in $\mathbb{P}^2_{\mathbb{C}}$ given by 

$$
2x+7y-5z=0
$$

and let $\mathcal{C}$ be the conic in $\mathbb{P}^2_{\mathbb{C}}$ given by

$$
2x^{2}-3xy+7y^{2}-5xz+11yz-8z^2=0.
$$

Note that the intersection of $L\cap\mathcal{C}$ with the line at infinity $z = 0$ is empty, since the system

$$
\begin{aligned}
2x^{2}-3xy+7y^{2}-5xz+11yz-8z^2 &=0,\\
2x+7y-5z&=0,\\
z&=0
\end{aligned}
$$

does not have solutions in $\mathbb{P}^2_{\mathbb{C}}$.
Thus, the intersection $L\cap\mathcal{C}$ is contained in the subset $U_z$ of $\mathbb{P}^2_{\mathbb{C}}$ given by $z\ne 0$.
Hence, to find $L\cap\mathcal{C}$, we have to solve the following system:

$$
\begin{aligned}
2x^{2}-3xy+7y^{2}-5xz+11yz-8z^2&=0,\\
2x+7y-5z&=0,\\
z&=1.
\end{aligned}
$$

This is easy to do with Python:

In [9]:
from sympy import *

x, y, z = symbols('x, y, z')

f = 2 * x ** 2 - 3 * x * y + 7 * y ** 2 - 5 * x * z + 11 * y * z - 8 * z ** 2
l = 2 * x + 7 * y - 5 * z

solution = solve((f, l, z - 1), x , y ,z)
P1 = solution[0]
P2 = solution[1]

print(P1)
print(P2)

(23/12 - sqrt(385)/12, 1/6 + sqrt(385)/42, 1)
(sqrt(385)/12 + 23/12, 1/6 - sqrt(385)/42, 1)


Running this code, we see that the intersection $L\cap\mathcal{C}$ consists of two points

$$
\Big[\frac{23\pm\sqrt{385}}{12}:\frac{7\mp\sqrt{385}}{42}:1\Big]=\Big[161\pm7\sqrt{385}:14\mp2\sqrt{385}:84\Big].
$$


### Exercise 5

Let $\mathcal{C}$ be the conic in Exercise 4.
Find the intersection $L\cap \mathcal{C}$, where $L$ is the line in $\mathbb{P}^2_{\mathbb{C}}$ containing the points $[-2:1:3]$ and $[1:2:3]$.

In [39]:
from sympy import  *
x, y, z = symbols('x, y, z')

P = (-2, 1, 3)
Q = (1, 2, 3)

M = Matrix([P,Q,[x,y,z]])
L = M.det()
print(L)

solution = solve((C5, L, z - 1), x , y ,z)
print(solution)


-3*x + 9*y - 5*z
[(-83/30, -11/30, 1), (-2/3, 1/3, 1)]


## 4. How many points determine a conic?

For two points $[a_{11}:a_{12}:a_{13}]\ne[a_{21}:a_{22}:a_{23}]$ in the projective plane $\mathbb{P}^2_{\mathbb{C}}$, the equation

$$
 \mathrm{det}\left(
\begin{matrix}
a_{11}& a_{12} & a_{13}\cr
a_{21}& a_{22} & a_{23} \cr
x& y & z \cr
\end{matrix}\right)=0
$$

determines the unique line in $\mathbb{P}^2_{\mathbb{C}}$ containing them if $[a_{11}:a_{12}:a_{13}]\ne[a_{21}:a_{22}:a_{23}]$.
This implies, in particular, that any three points $[a_{11}:a_{12}:a_{13}]$, $[a_{21}:a_{22}:a_{23}]$ and $[a_{31}:a_{32}:a_{33}]$
in $\mathbb{P}^2_{\mathbb{C}}$ are collinear (contained in one line) if and only if

$$
 \mathrm{det}\left(
\begin{matrix}
a_{11}& a_{12} & a_{13}\cr
a_{21}& a_{22} & a_{23}\cr
a_{31}& a_{32} & a_{33}\cr
\end{matrix}\right)=0.
$$

### Theorem

Suppose that every line in $\mathbb{P}^2_{\mathbb{C}}$ contains at most $3$ points among $P_1$, $P_2$, $P_3$, $P_4$, $P_5$.
Then there exists a unique conic $\mathcal{C}$ in $\mathbb{P}^2_{\mathbb{C}}$ that contains these five points.
Moreover, the conic $\mathcal{C}$ is irreducible if and only if every line in $\mathbb{P}^2_{\mathbb{C}}$
contains at most $2$ points among $P_1$, $P_2$, $P_3$, $P_4$, $P_5$.

Let's use Python to illustrate this result.
For example, if $P_1=[1:2:0]$, $P_2=[-1:2:1]$, $P_3=[3:1:2]$, $P_4=[2:5:0]$ and $P_5=[7:2:3]$, then the equation of the required conic is

$$
100x^2-90xy+20y^2-206xz-475yz+384z^2=0.
$$

This can be checked using the following Python script:

In [10]:
from sympy import *

x, y, z, A, B, C, D, E, F = symbols('x, y, z, A, B, C, D, E, F')

# Define a general conic

f = A * x ** 2 + B * x * y + C * y ** 2 + D * x * z + E * y * z + F * z ** 2

P1 = (1, 2, 0)
P2 = (-1, 2, 1)
P3 = (3, 1, 2)
P4 = (2, 5, 0)
P5 = (7, 2, 3)

# Set up a system of equations corresponding to the conditions that the points Pi lie in the conic

E1 = f.subs(((x, P1[0]), (y, P1[1]), (z, P1[2])))
E2 = f.subs(((x, P2[0]), (y, P2[1]), (z, P2[2])))
E3 = f.subs(((x, P3[0]), (y, P3[1]), (z, P3[2])))
E4 = f.subs(((x, P4[0]), (y, P4[1]), (z, P4[2])))
E5 = f.subs(((x, P5[0]), (y, P5[1]), (z, P5[2])))

M, b = linear_eq_to_matrix([E1, E2, E3, E4, E5], A, B, C, D, E, F)

# Print the rank of the system. If it is rank 5, then there is a unique solution (up to scaling). 

print(M.rank())

solution = solve((E1, E2, E3, E4, E5, F - 384), A, B, C, D, E, F)
coeff = (solution[A], solution[B], solution[C], solution[D], solution[E], solution[F])

# The solution corresponds to the coefficients of the desired polynomial f 

f = f.subs(((A, coeff[0]), (B, coeff[1]), (C, coeff[2]), (D, coeff[3]), (E, coeff[4]), (F, coeff[5])))

print(f)

5
100*x**2 - 90*x*y - 206*x*z + 20*y**2 - 475*y*z + 384*z**2


Let's rewrite the equation in matrix form:

$$
 \left(
\begin{array}{ccc} x & y & z \\
 \end{array}\right)\left( \begin{array}{ccc} 100 & -45 & -103 \\
 -45 & 20 & -\frac{475}{2} \\
 -103& -\frac{475}{2} & 384 \\
 \end{array}\right)\left(
\begin{array}{c} x \\
 y \\
 z \\
 \end{array}\right)=0.
$$

Then the determinant of the $3\times 3$ matrix in this equation equals $-8064030$, so that the conic we found is irreducible.
These computations can be done by running the following Python script:

In [11]:
from sympy import *

x, y, z = symbols('x,y,z')

# Define our conic

f = 100 * x ** 2 - 90 * x * y + 20 * y ** 2 - 206 * x * z - 475 * y * z + 384 * z ** 2

# Compute the hessian

M = hessian(f, (x, y, z))

# Print the determinant of the Hessian, scaled by 1/8.
# Recall, the determinant of the Hessian is 8 * the determinant of the coefficient matrix.

print(simplify(M.det()) / 8)

-8064030


### Exercise 6

Continuing the above example, show that no three points among $P_1$, $P_2$, $P_3$, $P_4$ and $P_5$ are collinear. Use this to give an alternative proof that the polynomial $100x^2-90xy+20y^2-206xz-475yz+384z^2$ is irreducible. 

Test your code by checking what happens when there are collinear triples. For example, consider adding a point ```P6 = (6, 4, 2)```. You should avoid deep loops by using  ```itertools.combinations``` from the ```itertools``` library.

In [59]:
from sympy import *
import itertools

P1 = (1, 2, 0)
P2 = (-1, 2, 1)
P3 = (3, 1, 2)
P4 = (2, 5, 0)
P5 = (7, 2, 3)

x,y,z = symbols('x,y,z')
f = 100*𝑥**2-90*𝑥*𝑦+20*𝑦**2-206*𝑥*𝑧-475*𝑦*𝑧+384*𝑧**2
P6 = (6, 4, 2)

points = [P1, P2, P3, P4, P5]
triples = []

for point0, point1, point2 in itertools.combinations(points, 3):
    
    # put three vector in a matrix M
    
    M = Matrix([point0, point1, point2])
    
    # check if determinate of M is zero, if yes stor triple in the list triples 
    
    if M.det() == 0:
        
        print(point0, point1, point2, ' are collinear')
        
        triples.append((point0, point1, point2))

if (len(triples)) == 0:
    print('no 3 points are colinear')

for point in points:
    V =  f.subs(x,point[0]).subs(y,point[1]).subs(z,point[2])
    if V== 0:
        print('The point', point, 'is on the curve')
print('since all 5 points are on the curve and they are all colinear, the curve is irreducible')

# add 6th point, so there should have colinear triple
points.append(P6)
for point0, point1, point2 in itertools.combinations(points, 3):
    # put three vector in a matrix M
    M = Matrix([point0, point1, point2])  
    # check if determinate of M is zero, if yes stor triple in the list triples 
    if M.det() == 0:
        print(point0, point1, point2, ' are collinear')
        triples.append((point0, point1, point2))
print('There are', len(triples), 'colinear triples')


no 3 point are colinear
The point (1, 2, 0) is on the curve
The point (-1, 2, 1) is on the curve
The point (3, 1, 2) is on the curve
The point (2, 5, 0) is on the curve
The point (7, 2, 3) is on the curve
since all 5 points are on the curve and they are all colinear, the curve is irreducible
(1, 2, 0) (7, 2, 3) (6, 4, 2)  are collinear
There are 1 colinear triples


### Exercise 7

Find a conic in $\mathbb{P}^2_{\mathbb{C}}$ that contains the points

$$
[3:4:5],\quad [5:12:13],\quad [8:15:17],\quad [7:24:25],\quad [20:21:29].
$$

Explain why this conic is unique. Why is the solution not a surprise?

In [63]:
from sympy import *
import itertools

x, y, z, A, B, C, D, E, F = symbols('x, y, z, A, B, C, D, E, F')

# Define a general conic

f = A * x ** 2 + B * x * y + C * y ** 2 + D * x * z + E * y * z + F * z ** 2

P1 = (3, 4, 5)
P2 = (5, 12, 13)
P3 = (8, 15, 17)
P4 = (7, 24, 25)
P5 = (20, 21, 29)

E1 = f.subs(((x,P1[0]), (y, P1[1]), (z, P1[2])))
E2 = f.subs(((x,P2[0]), (y, P2[1]), (z, P2[2])))
E3 = f.subs(((x,P3[0]), (y, P3[1]), (z, P3[2])))
E4 = f.subs(((x,P4[0]), (y, P4[1]), (z, P4[2])))
E5 = f.subs(((x,P5[0]), (y, P5[1]), (z, P5[2])))

solution = solve ((E1, E2, E3, E4, E5, A - 1), A, B, C, D, E, F)
coeff = (solution[A], solution[B], solution[C], solution[D], solution[E], solution[F])
f = f.subs(((A, coeff[0]), (B, coeff[1]), (C, coeff[2]), (D, coeff[3]), (E, coeff[4]), (F, coeff[5])))

print(f)
# add code here

6
x**2 + y**2 - z**2


### Exercise 8

Find a conic in $\mathbb{P}^2_{\mathbb{C}}$ that contains the points

$$
[0:1:2],\quad [0:3:4],\quad [0:5:6], \quad [7:8:0], \quad [9:10:0].
$$

Explain why this conic is unique. Why is the solution not a surprise?

In [79]:
from sympy import *

x, y, z, A, B, C, D, E, F = symbols('x, y, z, A, B, C, D, E, F')

# Define a general conic

f = A * x ** 2 + B * x * y + C * y ** 2 + D * x * z + E * y * z + F * z ** 2

P1 = (0, 1, 2)
P2 = (0, 3, 4)
P3 = (0, 5, 6)
P4 = (7, 8, 0)
P5 = (9, 10, 0)

E1 = f.subs(((x,P1[0]), (y, P1[1]), (z, P1[2])))
E2 = f.subs(((x,P2[0]), (y, P2[1]), (z, P2[2])))
E3 = f.subs(((x,P3[0]), (y, P3[1]), (z, P3[2])))
E4 = f.subs(((x,P4[0]), (y, P4[1]), (z, P4[2])))
E5 = f.subs(((x,P5[0]), (y, P5[1]), (z, P5[2])))

solution = solve ((E1, E2, E3, E4, E5, D - 1 ), A, B, C, D, E, F)
print (solution)

print('x * z = 0')

{A: -8*B/7 + 128*E/49 + 256*F/49, C: -2*E - 4*F, D: 1}
x * z = 0


## 5. Smooth and singular curves

Here we will define smooth conics, tangent lines and singular points in the setting of the projective plane. To do this, let's first consider an example. Let $\mathcal{C}_2$ be the conic given by

$$
13x^2-10xy+2y^2-44xz+16yz+15z^2=0,
$$

and let $L$ be the line given by

$$
12x-2y-86z=0.
$$

Then $L\cap\mathcal{C}_2=\{[11:23:1]\}$. This point lies in $U_z$. Let $\overline{\mathcal{C}}_2=U_z\cap\mathcal{C}_2$ and $\overline{L}=L\cap U_z$.
Then $\overline{\mathcal{C}}_2$ is a conic in $U_z$, which we identify with $\mathbb{C}^2$, given by

$$
 13\overline{x}^2-10\overline{x}\overline{y}+2\overline{y}^2-44\overline{x}+16\overline{y}+15=0,
$$

and $\overline{L}$ is a line in $U_z$ given by 

$$
12\overline{x}-2\overline{y}-86=0.
$$

We also have $[11:23:1]=(11,23)$ in coordinates $\overline{x}=\frac{x}{z}$ and $\overline{y}=\frac{y}{z}$.
Let $F(\overline{x},\overline{y})=13\overline{x}^2-10\overline{x}\overline{y}+2\overline{y}^2-44\overline{x}+16\overline{y}+15$. Then

$$
\begin{aligned}
&\frac{\partial F(\overline{x},\overline{y})}{\partial\overline{x}}=26\overline{x}-10\overline{y}-44,\\
&\frac{\partial F(\overline{x},\overline{y})}{\partial\overline{y}}=-10\overline{x}+4\overline{y}+16.
\end{aligned}
$$

Evaluating these partial derivatives at $(11,23)$,
we get $\frac{\partial F(\overline{x},\overline{y})}{\partial\overline{x}}(11,23)=12$
and $\frac{\partial F(\overline{x},\overline{y})}{\partial\overline{y}}(11,23)=-2$.
Hence, the equation of the tangent line to the curve $\overline{\mathcal{C}}_2$ at the point $(11,23)$ is given by

$$
12(\overline{x}-11)-2(\overline{y}-23)=0,
$$

which is exactly the equation of the line $\overline{L}$.
Thus, the line $\overline{L}$ is tangent to the conic $\overline{\mathcal{C}}_2$ at the point $(11,23)$,
so that $L$ is tangent to $\mathcal{C}_2$ at the point $[11:23:1]$.


Can we see this without using the subset $U_z$? Yes, we can.
Let 

$$
f(x,y,z)=13x^2-10xy+2y^2-44xz+16yz+15z^2.
$$

Then the polynomials $f(x,y,z)$ and $F(\overline{x},\overline{y})$ are related by 

$$
F(\overline{x},\overline{y})=f(\overline{x},\overline{y},1) \qquad \textrm{and} \qquad f(x,y,z)=z^2F\Big(\frac{x}{z},\frac{y}{z}\Big).
$$

Taking partial derivatives of the polynomial $f(x,y,z)$, we get

$$
\frac{\partial f(x,y,z)}{\partial x}=26x-10y-44z,\quad
\frac{\partial f(x,y,z)}{\partial y}=-10x-4y+16z,\quad
\frac{\partial f(x,y,z)}{\partial z}=-44x+16y+30z.
$$

Evaluating them at $[11:23:1]$, we get

$$
\frac{\partial f(x,y,z)}{\partial x}(11,23,1)=12,\quad
\frac{\partial f(x,y,z)}{\partial y}(11,23,1)=2,\quad
\frac{\partial f(x,y,z)}{\partial z}(11,23,1)=-86.
$$

Recall that $L$ is the tangent line to $\mathcal{C}_2$ at the point $[11:23:1]$, and $L$ is given by $12x-2y-86z=0$,
so that $L$ is given by

$$
\frac{\partial f(x,y,z)}{\partial x}(11,23,1)x+\frac{\partial f(x,y,z)}{\partial y}(11,23,1)y+\frac{\partial f(x,y,z)}{\partial z}(11,23,1)z=0.
$$

This is not a coincidence. The tangent line to $\mathcal{C}_2$ at every point $[a:b:c]\in\mathcal{C}_2$ is given by

$$
\frac{\partial f(x,y,z)}{\partial x}(a,b,c)x+\frac{\partial f(x,y,z)}{\partial y}(a,b,c)y+\frac{\partial f(x,y,z)}{\partial z}(a,b,c)z=0.
$$

Of course, this formula works for all curves in $\mathbb{P}^2_{\mathbb{C}}$.
It allows us to find tangent lines to curves of any degree.
However, there is a hidden problem: what if all partial derivatives are zero at some point?
Namely, if there is a point $[x:y:z]\in\mathcal{C}_2$ such that

$$
\frac{\partial f(x,y,z)}{\partial x}=0,\quad
\frac{\partial f(x,y,z)}{\partial y}=0,\quad
\frac{\partial f(x,y,z)}{\partial z}=0.
$$

In our example, this cannot happen, because the system of equations

$$
\begin{aligned}
26x-10y-44z&=0\\
-10x+4y+16z&=0\\
-44x+16y+30z&=0\\
\end{aligned}
$$

has only a trivial solution $x=y=z=0$, since

$$
\det\left(
\begin{array}{ccc}
26 & -10 & -44 \\
-10 & -4 & 16 \\
-44 & 16 & 30 \\
\end{array}\right)=-200\ne 0.
$$

This means that for every point $[a:b:c]\in\mathcal{C}_2$
there is a line in $\mathbb{P}^2_{\mathbb{C}}$ that is tangent to this conic at this point.
Note that the conic $\mathcal{C}_2$ is irreducible, because it is given by

$$
\left(
\begin{array}{ccc}
x & y & z \\
\end{array}\right)\left(\begin{array}{ccc}
26 & -10 & -44 \\
-10 & -4 & 16 \\
-44 & 16 & 30 \\
\end{array}\right)\left(\begin{array}{c}
x \\
y \\
z \\
\end{array}\right)=0,
$$

and we just computed the determinant of this $3\times 3$ matrix.

Now we are ready to give a definition in general. Let $\mathcal{C}_d$ be a curve in $\mathbb{P}^2_{\mathbb{C}}$ given by

$$
f_d(x,y,z)=0,
$$

where $f_d(x,y,z)$ is a homogeneous polynomial of degree $d\geqslant 1$. Let $[a:b:c]$ be a point in $\mathcal{C}_d$.
We say that $[a:b:c]$ is a *singular point* of the curve $\mathcal{C}_d$ if $[a:b:c]$ is a solution of the following system of equations:

$$
\frac{\partial f_d(x,y,z)}{\partial x}=0,\quad
\frac{\partial f_d(x,y,z)}{\partial y}=0,\quad
\frac{\partial f_d(x,y,z)}{\partial z}=0.
$$

Otherwise, we say that $[a:b:c]$ is a *smooth point* or a *non-singular point* of the curve $\mathcal{C}_d$
or that the curve $\mathcal{C}_d$ is *smooth* or *non-singular* at the point $[a:b:c]$.

If $[a:b:c]$ is a smooth point of the curve $\mathcal{C}_d$, then the equation

$$
\frac{\partial f_d(x,y,z)}{\partial x}(a,b,c)x+\frac{\partial f_d(x,y,z)}{\partial y}(a,b,c)y+\frac{\partial f_d(x,y,z)}{\partial z}(a,b,c)z=0
$$

defines a line in $\mathbb{P}^2_{\mathbb{C}}$. This is a *tangent* line to the curve $\mathcal{C}_d$ at the point $[a:b:c]$.
If all points of the curve $\mathcal{C}_d$ are smooth, we say that $\mathcal{C}_d$ is smooth.

### Exercise 9

Let $\mathcal{C}_2$ be the conic in $\mathbb{P}^2_{\mathbb{C}}$ given by

$$
x^2+xy-2y^2+3xz+3yz+z^2=0.
$$


- Verify that $\mathcal{C}_2$ does not have singular points.
- Show $\mathcal{C}_2$ contains $[-2:1:3]$.
- Find the equation of the line that is tangent to $\mathcal{C}_2$ at the point $[-2:1:3]$.
- Find the intersection of the conic $\mathcal{C}_2$ with this line.

In [91]:
from sympy import *

x, y, z = symbols ('x, y, z')
# irreducible ==? non singular ?
f =  x ** 2 + x * y - 2 * y ** 2 + 3 * x * z + 3 * y * z +  z ** 2

# Compute the hessian

M = hessian(f, (x, y, z))

if M.det() == 0:
    print('The curve has singular point')
else:
    print('The curve has no singular point')
    
V = f.subs(((x, -2),(y,1),(z,3)))
if V == 0:
    print('The points lies on curve')
else:
    print('The point doesnot lie on curve')

Cx = diff(f,x)
Cy = diff(f,y)
Cz = diff(f,z)

Vx = Cx.subs(((x, -2),(y,1),(z,3)))
Vy = Cy.subs(((x, -2),(y,1),(z,3)))
Vz = Cz.subs(((x, -2),(y,1),(z,3)))

L = Vx * x + Vy * y + Vz * z
print(L, '= 0')
solution = solve (( f, L, z-1 ), x, y, z)
print(solution) # exact = (-2,1,3)

The curve has no singular point
The points lies on curve
6*x + 3*y + 3*z = 0
[(-2/3, 1/3, 1)]


### Exercise 10

Find a smooth conic $\mathcal{C}_2$ in $\mathbb{P}^2_\mathbb{C}$ such that the following conditions are satisfied:
- the conic $\mathcal{C}_2$ contains the points $[0:0:1]$, $[0:1:0]$, $[1:0:0]$;
- the line $2x + y = 0$ is tangent to $\mathcal{C}_2$  at the point $[0:0:1]$;
- the line $y - z = 0$ is tangent to $\mathcal{C}_2$ at the point $[1:0:0]$.

In [101]:
from sympy import *

x, y, z, A, B, C, D, E, F = symbols('x, y, z, A, B, C, D, E, F')

# Define a general conic

f = A * x ** 2 + B * x * y + C * y ** 2 + D * x * z + E * y * z + F * z ** 2

Cx = diff(f, x)
Cy = diff(f, y)
Cz = diff(f, z)

P1 = (0, 0, 1)
P2 = (0, 1, 0)
P3 = (1, 0, 0)
E1 = f.subs(((x,0),(y,0),(z,1)))
E2 = f.subs(((x,0),(y,1),(z,0)))
E3 = f.subs(((x,1),(y,0),(z,0)))
E4 = Cx.subs(((x,0),(y,0),(z,1))) - 2 * Cy.subs(((x,0),(y,0),(z,1))) 
E5 = Cz.subs(((x,0),(y,0),(z,1)))
E6 = Cx.subs(((x,1),(y,0),(z,0)))
E7 = Cy.subs(((x,1),(y,0),(z,0))) -  Cz.subs(((x,1),(y,0),(z,0)))
sol = solve((E1, E2, E3, E4, E5, E6, E7, E - 1 ), A, B, C, D, E, F)
print(sol)
coeff = (sol[A], sol[B], sol[C], sol[D], sol[E], sol[F])

f = f.subs(((A, coeff[0]), (B, coeff[1]), (C, coeff[2]), (D, coeff[3]), (E, coeff[4]), (F, coeff[5])))
print(f) 

{F: 0, C: 0, A: 0, B: 2, D: 2, E: 1}
2*x*y + 2*x*z + y*z


Arguing as above, we see that a conic in $\mathbb{P}^2_{\mathbb{C}}$ is irreducible if and only if it is smooth.
In fact, one can show that every smooth plane curve is irreducible (see [Cheltsov, Lemma 2.42]).
However, there are irreducible singular cubic curves.
